File: HW1/Problem5/mergesort.js
"use strict";

const fs = require("fs");

const getRandomInt = (min, max) => {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min)) + min;
};

const genRandomNumbers = numToGen => {
    let numArray = [];

    for(let i = 0; i < numToGen; i++) {
        numArray.push(getRandomInt(0,10000));
    }

    return numArray;
};

const mergeSort = array => {
    const arraySize = array.length;

    if(arraySize === 1 || arraySize === 0) {
        return array;
    }

    const arraySplit = arraySize / 2;

    const n1 = mergeSort(array.slice(0, arraySplit));
    const n2 = mergeSort(array.slice(arraySplit));

    return merge(n1,n2);
};

const merge = (array1, array2) => {
    let newArray = [];

    while(array1.length != 0 && array2.length != 0) {
        if(array1[0] < array2[0]) {
            newArray.push(array1[0]);
            array1.splice(0,1);
        } else {
            newArray.push(array2[0]);
            array2.splice(0,1);
        }
    }

    if(array1.length === 0) {
        newArray = newArray.concat(array2);
    } else {
        newArray = newArray.concat(array1);        
    }

    return newArray;
};

const timeArray = [500,1000,1500,2000,2500,3000,3500];

fs.writeFileSync("mergetime.csv", "num,time\n");

timeArray.forEach(num => {
    for(let i = 0; i < 5; i++) {
        const randomArray = genRandomNumbers(num);
        const start = new Date();
        const hrstart = process.hrtime();
        const sortedArray = mergeSort(randomArray);
        const end = new Date() - start;
        const hrend = process.hrtime(hrstart);
        fs.appendFileSync("mergetime.csv", num + "," + end + "\n");
    }
})

File: HW1/Problem5/mergesort.py
from random import randint
import timeit

def genRandomNumbers(numToGen):
	numArray = []

	for x in xrange(0,numToGen):
		numArray.append(randint(0,10000))

	return numArray

def mergeSort(array):
	arraySize = len(array)

	if arraySize == 1 or arraySize == 0:
		return array

	arraySplit = arraySize / 2

	n1 = mergeSort(array[:arraySplit])
	n2 = mergeSort(array[arraySplit:])
	
	return merge(n1, n2)

def merge(array1, array2):
	newArray = []

	while len(array1) != 0 and len(array2) != 0:

		if array1[0] < array2[0]:
			newArray.append(array1.pop(0))
		else:
			newArray.append(array2.pop(0))

	if len(array1) == 0:
		newArray += array2
	else:
		newArray += array1

	return newArray

timeArray = (1000,2000,5000,10000,20000,30000,35000)

outFile = open("mergetime.csv", "w")

outFile.write("num,time\n")

for num in timeArray:
	for x in xrange(0,5):
		randomArray = genRandomNumbers(num)
		start_time = timeit.default_timer()
		sortedArray = mergeSort(randomArray)
		elapsed = timeit.default_timer() - start_time

		outFile.write(str(num) + "," + str(elapsed) + '\n')

outFile.close()
